src/core/lib/surface/call.cc

```diff
@@ -1674,135 +1631,241 @@ static grpc_call_error call_start_batch(grpc_call* call, const grpc_op* ops,
+    grpc_call* call) {
+  return grpc_core::Call::FromC(call)->test_only_compression_algorithm();
+}
+
+uint32_t grpc_call_test_only_get_message_flags(grpc_call* call) {
+  return grpc_core::Call::FromC(call)->test_only_message_flags();
+}
+
+uint32_t grpc_call_test_only_get_encodings_accepted_by_peer(grpc_call* call) {
+  return grpc_core::Call::FromC(call)->test_only_encodings_accepted_by_peer();
+}
+
+grpc_core::Arena* grpc_call_get_arena(grpc_call* call) {
+  return grpc_core::Call::FromC(call)->arena();
+}
+
+grpc_call_stack* grpc_call_get_call_stack(grpc_call* call) {
+  return grpc_core::Call::FromC(call)->call_stack();
+}
+
 grpc_call_error grpc_call_start_batch(grpc_call* call, const grpc_op* ops,
                                       size_t nops, void* tag, void* reserved) {
-  grpc_call_error err;
-
   GRPC_API_TRACE(
       "grpc_call_start_batch(call=%p, ops=%p, nops=%lu, tag=%p, "
       "reserved=%p)",
       5, (call, ops, (unsigned long)nops, tag, reserved));
 
   if (reserved != nullptr) {
-    err = GRPC_CALL_ERROR;
+    return GRPC_CALL_ERROR;
   } else {
     grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
-    err = call_start_batch(call, ops, nops, tag, 0);
+    return grpc_core::Call::FromC(call)->StartBatch(ops, nops, tag, false);
   }
-
-  return err;
 }
 
 grpc_call_error grpc_call_start_batch_and_execute(grpc_call* call,
                                                   const grpc_op* ops,
                                                   size_t nops,
                                                   grpc_closure* closure) {
-  return call_start_batch(call, ops, nops, closure, 1);
+  return grpc_core::Call::FromC(call)->StartBatch(ops, nops, closure, true);
 }
 
 void grpc_call_context_set(grpc_call* call, grpc_context_index elem,
                            void* value, void (*destroy)(void* value)) {
-  if (call->context[elem].destroy) {
-    call->context[elem].destroy(call->context[elem].value);
-  }
-  call->context[elem].value = value;
-  call->context[elem].destroy = destroy;
+  return grpc_core::Call::FromC(call)->ContextSet(elem, value, destroy);
 }
 
 void* grpc_call_context_get(grpc_call* call, grpc_context_index elem) {
-  return call->context[elem].value;
+  return grpc_core::Call::FromC(call)->ContextGet(elem);
 }
 
-uint8_t grpc_call_is_client(grpc_call* call) { return call->is_client; }
+uint8_t grpc_call_is_client(grpc_call* call) {
+  return grpc_core::Call::FromC(call)->is_client();
+}
 
 grpc_compression_algorithm grpc_call_compression_for_level(
     grpc_call* call, grpc_compression_level level) {
-  return call->encodings_accepted_by_peer.CompressionAlgorithmForLevel(level);
+  return grpc_core::Call::FromC(call)->compression_for_level(level);
 }
 
 bool grpc_call_is_trailers_only(const grpc_call* call) {
-  bool result = call->is_trailers_only;
-  GPR_DEBUG_ASSERT(!result || call->recv_initial_metadata.empty());
```

Comment:
can we keep this assert for now? Alternatively, we make progress on https://github.com/grpc/grpc/pull/28847/files

Reply:
It's still there, but it moved into `Call::is_trailers_only()`